src/common/signalk/SKNMEA2000Parser.cpp

```diff
@@ -80,42 +128,207 @@ const SKUpdate& SKNMEA2000Parser::parse128259(const SKSourceInput& input, const
+        // Ground Wind
+        update->setEnvironmentWindSpeedTrue(windSpeed);
+        // Ground Wind +/- starboard/port
+        if (windAngle >M_PI) windAngle *= -1;
+        update->setEnvironmentWindAngleTrueGround(windAngle);
+      }
+      else if (windReference == N2kWind_True_water) {
+        // TWS (water refered) True "Sailing" Wind
+        update->setEnvironmentWindSpeedTrue(windSpeed);
+        // TWA (water refered) +/- starboard/port
+        if (windAngle >M_PI) windAngle *= -1;
+        update->setEnvironmentWindAngleTrueWater(windAngle);
+      }
     }
 
     _sku = update;
     return *_sku;
   }
   else {
-    INFO("Unable to parse N2kMsg with PGN %i", msg.PGN);
+    DEBUG("Unable to parse NMEA2000 with PGN %i", msg.PGN);
     return _invalidSku;
   }
 }
+
+//  ***********************************************
+//   PGN 127245 Rudder
+//        →  sid
+//        →  RudderPosition [rad]
+// *  ********************************************** */
+const SKUpdate& SKNMEA2000Parser::parse127245(const SKSourceInput& input, const tN2kMsg& msg, const SKTime& timestamp) {
+  unsigned char instance;
+  tN2kRudderDirectionOrder rudderDirectionOrder;
+  double rudderPosition = N2kDoubleNA;
+  double angleOrder = N2kDoubleNA;
+
+  if (ParseN2kRudder(msg,rudderPosition,instance,rudderDirectionOrder,angleOrder)) {
+    // validation check max +/- 45°
+    if (!N2kIsNA(rudderPosition) && (rudderPosition <= M_PI_4) && (rudderPosition >= M_PI_4) ) {
+      SKUpdateStatic<1> *update = new SKUpdateStatic<1>();
+      update->setTimestamp(timestamp);
+
+      SKSource source = SKSource::sourceForNMEA2000(input, msg.PGN, msg.Priority, msg.Source);
+      update->setSource(source);
+      // -> Current rudder angle, +ve is rudder to Starboard
+      update->setSteeringRudderAngle(rudderPosition);
+
+      _sku = update;
+      return *_sku;
+    }
+  }
+  DEBUG("Unable to parse NMEA2000 with PGN %i", msg.PGN);
+  return _invalidSku;
+}
+// *****************************************************************************
+//    PGN 127250 VESSEL HEADING RAPID
+//    Heading sensor value with a flag for True or Magnetic.
+//    Heading               Heading in radians
+//      - Deviation         Magnetic deviation in radians. Use N2kDoubleNA for undefined value.
+//      - Variation         Magnetic variation in radians. Use N2kDoubleNA for undefined value.
+//      tN2kHeadingReference
+//        N2khr_true=0,
+//        N2khr_magnetic=1
+// *****************************************************************************
+const SKUpdate& SKNMEA2000Parser::parse127250(const SKSourceInput& input, const tN2kMsg& msg, const SKTime& timestamp) {
+  unsigned char sid;
+  tN2kHeadingReference headingReference;
+  double heading = N2kDoubleNA;
+  double deviation = N2kDoubleNA;
+  double variation = N2kDoubleNA;
+
+  if (ParseN2kHeading(msg,sid,heading,deviation,variation,headingReference)) {
+    if (!N2kIsNA(heading) && heading >= 0 && heading <= 2 * M_PI) {
```

Comment:
Are you sure we need to check that heading is between 0 and 2 PI? Other values are valid too but we could just normalize it to make sure whatever we publish is between 0 and 2 pi.

Reply:
I do not know if heading of more of 360° is allowed.
In AIS it would be treated as not valid.

Reply:
Ok.  As discussed let's keep it like this.